solutions/Z3.Linq/ExpressionVisitor.cs

```
@@ -0,0 +1,430 @@
+        var hierarchy = new List<MemberExpression>();
+        var mExp = member;
+        hierarchy.Add(mExp);
+
+        while (mExp.Expression is MemberExpression parent)
+        {
+            mExp = parent;
+            hierarchy.Add(parent);
+        }
+
+        hierarchy.Reverse();
+
+        var topMember = hierarchy.First();
+
+        if (topMember.Expression != param)
+        {
+            if ((topMember.Expression is ConstantExpression expression))
+            {
+                // We only ever get here if SimplifyLambda is set to false, otherwise partial evaluation does it earlier
+                var target = expression.Value;
+                var hierarchyIdx = 0;
+                object? val = target;
+
+                while (hierarchyIdx < hierarchy.Count)
+                {
+                    var currentMember = hierarchy[hierarchyIdx].Member;
+
+                    switch (currentMember.MemberType)
+                    {
+                        case MemberTypes.Property:
+                            var property = (PropertyInfo)currentMember;
+                            val = property.GetValue(target);
+                            break;
+                        case MemberTypes.Field:
+                            var field = (FieldInfo)currentMember;
+                            val = field.GetValue(target);
+                            break;
+                        default:
+                            throw new NotSupportedException($"Unsupported constant {target} .");
+                    }
+                        
+                    hierarchyIdx++;
+                }
+
+                if (val != null)
+                {
+                    return VisitConstantValue(context, val);
+                }
+
+                throw new NotSupportedException($"Could not reduce expression {topMember.Expression}");
+            }
+            else
+            {
+                //Debugger.Break(); 
```

Eh? do we need to throw an exception? Or should this:

```cs
topMember.Expression is ConstantExpression expression
```

just be a plain cast? If it's a "should never happen" for it to be anything other than a `ConstantExpression` then this can be a cast. If it's a "user should never supply input in a form that makes this happen" then we should throw some more informative exception.